File: test/VbNet.LanguageServer.Tests/Integration/SemanticTokensIntegrationTests.cs
using Microsoft.Extensions.Logging.Abstractions;
using VbNet.LanguageServer.Protocol;
using VbNet.LanguageServer.Services;
using VbNet.LanguageServer.Workspace;
using Xunit;

namespace VbNet.LanguageServer.Tests.Integration;

/// <summary>
/// Integration tests for SemanticTokensService with real VB.NET projects.
/// </summary>
[Collection("MSBuild")]
public class SemanticTokensIntegrationTests : IAsyncLifetime
{
    private readonly WorkspaceManager _workspaceManager;
    private readonly DocumentManager _documentManager;
    private readonly SemanticTokensService _semanticTokensService;

    private static readonly string TestProjectsRoot = GetTestProjectsRoot();

    public SemanticTokensIntegrationTests()
    {
        _workspaceManager = new WorkspaceManager(NullLogger<WorkspaceManager>.Instance);
        _documentManager = new DocumentManager(_workspaceManager, NullLogger<DocumentManager>.Instance);
        _semanticTokensService = new SemanticTokensService(
            _workspaceManager,
            _documentManager,
            NullLogger<SemanticTokensService>.Instance);
    }

    private static string GetTestProjectsRoot()
    {
        var assemblyLocation = typeof(SemanticTokensIntegrationTests).Assembly.Location;
        var assemblyDir = Path.GetDirectoryName(assemblyLocation)!;
        return Path.GetFullPath(Path.Combine(assemblyDir, "..", "..", "..", "..", "TestProjects"));
    }

    public Task InitializeAsync()
    {
        _workspaceManager.Initialize();
        return Task.CompletedTask;
    }

    public async Task DisposeAsync()
    {
        await _workspaceManager.DisposeAsync();
    }

    [Fact]
    public async Task GetSemanticTokensAsync_ReturnsTokenData()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "SmallProject", "SmallProject.vbproj");
        var helperPath = Path.Combine(TestProjectsRoot, "SmallProject", "Helper.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var helperUri = new Uri(helperPath).ToString();
        var text = await File.ReadAllTextAsync(helperPath);

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = helperUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        var result = await _semanticTokensService.GetSemanticTokensAsync(new SemanticTokensParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = helperUri }
        }, CancellationToken.None);

        Assert.NotNull(result);
        Assert.NotEmpty(result.Data);
        Assert.True(result.Data.Length % 5 == 0);

        var legend = SemanticTokensService.GetLegend();
        for (var i = 0; i < result.Data.Length; i += 5)
        {
            var tokenType = result.Data[i + 3];
            Assert.True(tokenType < legend.TokenTypes.Length);
        }
    }
}

File: test/VbNet.LanguageServer.Tests.Vb/Integration/SemanticTokensIntegrationTests.vb
Imports System
Imports System.IO
Imports System.Threading
Imports System.Threading.Tasks
Imports Microsoft.Extensions.Logging.Abstractions
Imports VbNet.LanguageServer.Protocol
Imports VbNet.LanguageServer.Services
Imports VbNet.LanguageServer.Workspace
Imports Xunit

Namespace VbNet.LanguageServer.Tests.Integration

    ''' <summary>
    ''' Integration tests for SemanticTokensService with real VB.NET projects.
    ''' </summary>
    <Collection("MSBuild")>
    Public Class SemanticTokensIntegrationTests
        Implements IAsyncLifetime

        Private ReadOnly _workspaceManager As WorkspaceManager
        Private ReadOnly _documentManager As DocumentManager
        Private ReadOnly _semanticTokensService As SemanticTokensService

        Private Shared ReadOnly TestProjectsRoot As String = GetTestProjectsRoot()

        Public Sub New()
            _workspaceManager = New WorkspaceManager(NullLogger(Of WorkspaceManager).Instance)
            _documentManager = New DocumentManager(_workspaceManager, NullLogger(Of DocumentManager).Instance)
            _semanticTokensService = New SemanticTokensService(
                _workspaceManager,
                _documentManager,
                NullLogger(Of SemanticTokensService).Instance)
        End Sub

        Private Shared Function GetTestProjectsRoot() As String
            Dim assemblyLocation = GetType(SemanticTokensIntegrationTests).Assembly.Location
            Dim assemblyDir = Path.GetDirectoryName(assemblyLocation)
            Return Path.GetFullPath(Path.Combine(assemblyDir, "..", "..", "..", "..", "TestProjects"))
        End Function

        Public Function InitializeAsync() As Task Implements IAsyncLifetime.InitializeAsync
            _workspaceManager.Initialize()
            Return Task.CompletedTask
        End Function

        Public Async Function DisposeAsync() As Task Implements IAsyncLifetime.DisposeAsync
            Await _workspaceManager.DisposeAsync()
        End Function

        <Fact>
        Public Async Function GetSemanticTokensAsync_ReturnsTokenData() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "SmallProject", "SmallProject.vbproj")
            Dim helperPath = Path.Combine(TestProjectsRoot, "SmallProject", "Helper.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim helperUri = New Uri(helperPath).ToString()
            Dim text = Await File.ReadAllTextAsync(helperPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = helperUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim result = Await _semanticTokensService.GetSemanticTokensAsync(New SemanticTokensParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = helperUri}
            }, CancellationToken.None)

            Assert.NotNull(result)
            Assert.NotEmpty(result.Data)
            Assert.True(result.Data.Length Mod 5 = 0)

            Dim legend = SemanticTokensService.GetLegend()
            Dim i = 0
            While i < result.Data.Length
                Dim tokenType = result.Data(i + 3)
                Assert.True(tokenType < legend.TokenTypes.Length)
                i += 5
            End While
        End Function
    End Class

End Namespace
